test/unit/org/apache/cassandra/index/sai/SAITester.java

```diff
@@ -334,15 +365,49 @@ protected boolean verifyChecksum(IndexContext context)
 
         for (SSTableReader sstable : cfs.getLiveSSTables())
         {
-            IndexDescriptor indexDescriptor = IndexDescriptor.createFrom(sstable);
+            IndexDescriptor indexDescriptor = loadDescriptor(sstable, cfs);
             if (indexDescriptor.isIndexEmpty(context))
                 continue;
-            if (!indexDescriptor.validatePerSSTableComponentsChecksum() || !indexDescriptor.validatePerIndexComponentsChecksum(context))
+            if (!indexDescriptor.perSSTableComponents().validateComponents(sstable, cfs.getTracker(), true)
+                || !indexDescriptor.perIndexComponents(context).validateComponents(sstable, cfs.getTracker(), true))
                 return false;
         }
         return true;
     }
 
+    protected void verifySAIVersionInUse(Version expectedVersion, IndexContext... contexts)
+    {
+        ColumnFamilyStore cfs = Keyspace.open(KEYSPACE).getColumnFamilyStore(currentTable());
+        StorageAttachedIndexGroup group = StorageAttachedIndexGroup.getIndexGroup(cfs);
+
+        for (SSTableReader sstable : cfs.getLiveSSTables())
+        {
+            IndexDescriptor indexDescriptor = loadDescriptor(sstable, cfs);
+
+            assertEquals(indexDescriptor.perSSTableComponents().version(), expectedVersion);
+            SSTableContext ssTableContext = group.sstableContextManager().getContext(sstable);
+            // This is to make sure the context uses the actual files we think
+            assertEquals(ssTableContext.usedPerSSTableComponents().version(), expectedVersion);
+
+            for (IndexContext indexContext : contexts)
+            {
+                assertEquals(indexDescriptor.perIndexComponents(indexContext).version(), expectedVersion);
+
+                for (SSTableIndex sstableIndex : indexContext.getView())
+                {
+                    if (sstableIndex.isEmpty())
+                        continue;
+
+                    // Make sure the index does use the context it should (for which we already checked the version),
+                    // but note that `SSTableIndex` uses a "sharedCopy()" of the context, so we check equality of the
+                    // pk map factory (which should be same between original and a copy).
+                    assertEquals(sstableIndex.usedPerIndexComponents(), expectedVersion);
```

Comment:
Good catch. Especially since the test relying on this was passing, which was suss. Turns out the test in question was not passing the proper `IndexContext` values, so the `View` was always empty and this was never tested. Fixed that as well and double-checked this now assert things correctly.